rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,105 @@
+[design]: #detailed-design
+
+Standardize `passthru.tests.<name>` and `passthru.nixosTests.<name>` as a mechanism of 
+more expensive but automatic testing for nixpkgs. As well as encourage the usage of
+`checkPhase` or `installCheckPhase` when packaging within nixpkgs.
+
+Usage for `passthru.nixosTests.<name>`
+- Reserved for tests utilitizing the nixosTest utilties.
+  - Generally these are more resource intensive, and may require additional system features
+  such as kvm
+
+Usage for `passthru.tests.<name>`:
+- Running tests which include downstream dependencies.
+  - This avoids cyclic dependency issues for test suites.
+- Running lengthy or more resource expensive tests.
+  - There should be a priority on making package builds as short as possible.
+  - This reduces the amount of compute required for everyone reviewing, building, or iterating on packages.
+- Referencing downstream dependencies which are most likely to experience regressions.
+  - Most applicable to [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88),
+as this will help define what breakages a pull request author should take ownership.
+
+Usage for mkDerivation's `checkPhase`:
+- Quick "cheap" tests, which run units tests and maybe some addtional scenarios.
+- Since this contributes to the "build time" of a package, there should be some
+emphasis on ensuring this phase isn't bloated.
+
+Usage for mkDerivations `installCheckPhase`:
+- A quick trivial example (e.g. `<command> --help`) to demonstrate that one (or more)
+of the programs were linked correctly.
+- Assert behavior post installation (e.g. python's native extensions only get installed
+and are not present in a build directory)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None? This is opt-in behavior for package maintainers.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Continue to use current ad-hoc conventions.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+How far should testing go?
+- What consistitutes that "enough testing" was done to a package before a change was merged?
+
+Should `<package>.passthru.tests` be flat?
+For packages which have extremes in resource usage when testing (e.g. pytorch), it may
+be beneficial to have additional structure for the tests to denote expectations of resources
+and ownership of testing for upstream packages.
+
+# Future work
+[future]: #future-work
+
+Problem with onboarding more test to aspects of nixpkgs CI and processes is the increased
+need of compute, storage, and ram resources. Therefore, consideration of future work should
+take into consideration how much testing is feasible for a given change.
+
+Onboarding of CI tools to support testing paradigms:
+- ofborg
+  - Testing of `<package>.passthru.tests` is already done.
+  - Testing of downstream dependencies and their tests when minimal (e.g. <10 rebuilds?)
+- hydra
+  - Allow for derivations exposed to hydraJobs to also probe for `<drv>.passthru.tests`?
```

Comment:
This could be implemented in `release.nix` files.

Reply:
I removed this line, and added it to unresolved section, as it's more of a hydra concern than anything.

Also, the ultimate goal of hydra is to provide channel updates for nixpkgs, in which these tests are more "nice to have" than anything else. 

Also, I'm not sure if the official hydra should be building all of the tests, if the tests are only meant to influence whether  a change should be merged. The official hydra could be populating the cache with a lot of never-to-be-used-again builds.